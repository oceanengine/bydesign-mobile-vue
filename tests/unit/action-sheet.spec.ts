import { mount, later } from '../index';
import { ActionSheet as ActionSheetNative } from '@src';
const ActionSheet = ActionSheetNative as any;

later;
describe('ActionSheet', () => {
  test('callback events', async () => {
    const callback = jest.fn();
    const onClose = jest.fn();
    const onCancel = jest.fn();
    const onSelect = jest.fn();

    const actions = [
      { name: 'Option', callback },
      { name: 'Option', disabled: true },
      { name: 'Option', loading: true },
      { name: 'Option', subname: 'Subname' },
    ];

    const wrapper = mount(ActionSheet, {
      propsData: {
        value: true,
        actions,
        cancelText: 'Cancel',
      },
      context: {
        on: {
          close: onClose,
          cancel: onCancel,
          select: onSelect,
        },
      },
    });
    const options = wrapper.findAll('.byted-action-sheet__item');

    options.at(0).trigger('click');
    options.at(1).trigger('click');
    wrapper.find('.byted-action-sheet__cancel').trigger('click');

    await wrapper.setProps({ value: false });

    expect(callback).toHaveBeenCalled();
    expect(onCancel).toHaveBeenCalled();
    expect(onSelect).toHaveBeenCalledWith(actions[0], 0);
    expect(wrapper).toMatchSnapshot();
    expect(onClose).toHaveBeenCalled();
  });
  test('click overlay and close', async () => {
    const onOpen = jest.fn();
    const onClickOverlay = jest.fn();
    const div = document.createElement('div');

    const wrapper = mount({
      template: `
      <div>
        <action-sheet
          :value="show"
          :get-container="getContainer"
          @open="onOpen"
          @click-overlay="onClickOverlay"
        />
      </div>
    `,
      components: {
        ActionSheet,
      },
      data() {
        return {
          show: false,
          getContainer: () => div,
        };
      },
      methods: {
        onOpen,
        onClickOverlay,
      },
    }) as any;
    await wrapper.setData({ show: true });
    await later();
    (div.querySelector('.byted-overlay') as HTMLElement).click();

    expect(onOpen).toHaveBeenCalledTimes(1);
    expect(onClickOverlay).toHaveBeenCalledTimes(1);
  });
  test('disable lazy-render', () => {
    const wrapper = mount(ActionSheet, {
      propsData: {
        lazyRender: false,
        actions: [{ name: 'Option' }, { name: 'Option' }],
        cancelText: 'Cancel',
      },
    });

    expect(wrapper).toMatchSnapshot();
  });
  test('render title and default slot', () => {
    const wrapper = mount(ActionSheet, {
      propsData: {
        value: true,
        title: 'Title',
      },
      scopedSlots: {
        default() {
          return 'Default';
        },
      },
    });

    expect(wrapper).toMatchSnapshot();
  });
  test('get container', () => {
    const wrapper = mount(ActionSheet, {
      propsData: {
        value: true,
        getContainer: 'body',
      },
    });

    expect(wrapper.vm.$el.parentNode).toEqual(document.body);
  });
  test('extra slot', () => {
    const wrapper = mount(ActionSheet, {
      propsData: {
        value: true,
        actions: [{ name: 'Option' }],
      },
      scopedSlots: {
        extra: () => 'Custom extra',
      },
    });

    expect(wrapper).toMatchSnapshot();
  });
});
