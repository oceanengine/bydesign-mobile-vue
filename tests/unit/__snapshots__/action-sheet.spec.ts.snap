// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionSheet callback events 1`] = `
<transition-stub name="byted-popup-slide-bottom" actions="[object Object],[object Object],[object Object],[object Object]" canceltext="Cancel" class="byted-action-sheet" style="z-index: 2001;">
  <div class="byted-popup byted-popup--bottom" style="display: none;">
    <div class="byted-action-sheet__item">
      <div class="byted-action-sheet__option-content"><span class="byted-action-sheet__name">Option</span></div>
    </div>
    <div class="byted-action-sheet__item byted-action-sheet__item--disabled">
      <div class="byted-action-sheet__option-content"><span class="byted-action-sheet__name">Option</span></div>
    </div>
    <div class="byted-action-sheet__item byted-action-sheet__item--disabled">
      <div class="byted-loading"><i class="byted-loading__icon byted-loading-circle" style="width: 1rem; height: 1rem;">test-file-stub</i></div>
    </div>
    <div class="byted-action-sheet__item">
      <div class="byted-action-sheet__option-content byted-action-sheet__option-content--subname"><span class="byted-action-sheet__name">Option</span><span class="byted-action-sheet__subname">Subname</span></div>
    </div>
    <div class="byted-action-sheet__cancel">
      <div class="byted-action-sheet__cancel__text">Cancel</div>
    </div>
  </div>
</transition-stub>
`;

exports[`ActionSheet disable lazy-render 1`] = `
<transition-stub name="byted-popup-slide-bottom" actions="[object Object],[object Object]" canceltext="Cancel" class="byted-action-sheet">
  <div class="byted-popup byted-popup--bottom" style="display: none;">
    <div class="byted-action-sheet__item">
      <div class="byted-action-sheet__option-content"><span class="byted-action-sheet__name">Option</span></div>
    </div>
    <div class="byted-action-sheet__item">
      <div class="byted-action-sheet__option-content"><span class="byted-action-sheet__name">Option</span></div>
    </div>
    <div class="byted-action-sheet__cancel">
      <div class="byted-action-sheet__cancel__text">Cancel</div>
    </div>
  </div>
</transition-stub>
`;

exports[`ActionSheet extra slot 1`] = `
<transition-stub name="byted-popup-slide-bottom" actions="[object Object]" class="byted-action-sheet">
  <div class="byted-popup byted-popup--bottom">
    <div class="byted-action-sheet__item">
      <div class="byted-action-sheet__option-content"><span class="byted-action-sheet__name">Option</span></div>
      <div class="byted-action-sheet__option-extra">Custom extra</div>
    </div>
  </div>
</transition-stub>
`;

exports[`ActionSheet render title and default slot 1`] = `
<transition-stub name="byted-popup-slide-bottom" title="Title" class="byted-action-sheet">
  <div class="byted-popup byted-popup--bottom">
    <div class="byted-action-sheet__header">Title</div>
    <div class="byted-action-sheet__content">Default</div>
  </div>
</transition-stub>
`;
